### PR TITLE
Bug fixes + Expanded support for certain methods to support CP2K on Cori

### PR DESCRIPTION
**Pull request by ghost**
Changes:
- Added support for MPI_Sendrecv_replace method (+ tests for this method and MPI_Sendrecv)
- Added support for MPI_Type_struct (+ macros to build on Cori/CentOS)
- Fixed segfault caused by not checkpointing xpmem library
- Fixed crash caused by uninitialized return code for MPI collective barriers called with 0 requests
- Added MPI_Error_string to Fortran wrappers

These are all of the changes required to run CP2K on Cori, except for adding MPI_Dims_create to the Fortran wrappers  (which should not be done until the cartesian support is ready to be merged)

### CODE REVIEW

**Comment by chirag-singh-memverge** [comment]
@JainTwinkle @gc00 Please review

File: mpi-proxy-split/mpi-wrappers/mpi_p2p_wrappers.cpp [hunk 1]
```diff
@@ -247,23 +247,22 @@ USER_DEFINED_WRAPPER(int, Sendrecv, (const void *) sendbuf, (int) sendcount,
...
 
-// FIXME: Move this to mpi_collective_wrappers.cpp and reimplement
 USER_DEFINED_WRAPPER(int, Sendrecv_replace, (void *) buf, (int) count,
                      (MPI_Datatype) datatype, (int) dest,
                      (int) sendtag, (int) source,
                      (int) recvtag, (MPI_Comm) comm, (MPI_Status *) status)
 {
-  JASSERT(false).Text("MPI_Sendrecv_replace is not supported");
+
+  //Send first
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi-wrappers/mpi_p2p_wrappers.cpp:256]: Minor: space after `//`

File: mpi-proxy-split/mpi-wrappers/mpi_p2p_wrappers.cpp [hunk 2]
```diff
@@ -247,23 +247,22 @@ USER_DEFINED_WRAPPER(int, Sendrecv, (const void *) sendbuf, (int) sendcount,
...
-                                       dest, sendtag, source, recvtag,
-                                       realComm, status);
-  RETURN_TO_UPPER_HALF();
-  DMTCP_PLUGIN_ENABLE_CKPT();
+  retval = MPI_Send(buf, count, datatype, dest, sendtag, comm);
+  if (retval != MPI_SUCCESS) {
+    return retval;
+  }
+
+  //Recv using asynchronous to fill status struct
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi-wrappers/mpi_p2p_wrappers.cpp:263]: Minor: space after `//`
Please update the comment. It's no longer asynchronous.

File: mpi-proxy-split/mpi-wrappers/mpi_p2p_wrappers.cpp [hunk 3]
```diff
@@ -247,23 +247,22 @@ USER_DEFINED_WRAPPER(int, Sendrecv, (const void *) sendbuf, (int) sendcount,
...
+  retval = MPI_Recv(buf, count, datatype, source, recvtag, comm, status);
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi-wrappers/mpi_p2p_wrappers.cpp:264]: @xuyao0127,
Chirag was calling  `MPI_Irecv + MPI_Waitall` here earlier (for the mpi status). I then suggested calling `MPI_recv` that takes status as an arg.
Would you suggest one over another (considering the drain during the checkpoint)?

File: mpi-proxy-split/test/sendrecv_replace_test.c [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#include <unistd.h>
+#include <limits.h>
+#include <time.h>
+#include <getopt.h>
+#include <string.h>
+
+#define BUFFER_SIZE 3
+
+int main(int argc, char** argv) {
+  //Parse runtime argument
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_replace_test.c:21]: Minor: space after `//`

File: mpi-proxy-split/test/sendrecv_replace_test.c [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        return 1;
+    }
+  }
+
+  int buf[BUFFER_SIZE];
+  int exp[BUFFER_SIZE];
+
+  // Initialize the MPI environment
+  MPI_Init(NULL, NULL);
+  // Find out rank, size
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_replace_test.c:46]: rank and size

File: mpi-proxy-split/test/sendrecv_replace_test.c [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  int rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_replace_test.c:48]: Let's check the return value for both MPI calls.

File: mpi-proxy-split/test/sendrecv_replace_test.c [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  int world_size;
+  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+
+  if(rank == 0){
+    printf("Running test for %d iterations\n", max_iterations);
+  }
+
+  // We are assuming 2 processes for this task
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_replace_test.c:56]: 2 or 3?

File: mpi-proxy-split/test/sendrecv_replace_test.c [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  for(int iterations = 0; iterations < max_iterations; iterations++){
+    for(int i = 0; i < BUFFER_SIZE; i++){
+      buf[i] = iterations + rank + i;
+      exp[i] = iterations + ((rank+1)%3) + i;
+    }
+    int dst = (rank+1) % 3;
+    int src = (rank+2) % 3;
+    int tag = 123;
+    MPI_Sendrecv_replace(buf, BUFFER_SIZE, MPI_INT, src, iterations+tag,
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_replace_test.c:71]: Let's check the return value for error!

File: mpi-proxy-split/test/sendrecv_test.c [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+#include <unistd.h>
+#include <limits.h>
+#include <time.h>
+#include <getopt.h>
+#include <string.h>
+
+#define BUFFER_SIZE 3
+
+int main(int argc, char** argv) {
+  //Parse runtime argument
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_test.c:21]: Minor: space after `//`

File: mpi-proxy-split/test/sendrecv_test.c [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  int rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  int world_size;
+  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+
+  if(rank == 0){
+    printf("Running test for %d iterations\n", max_iterations);
+  }
+
+  // We are assuming 2 processes for this task
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/sendrecv_test.c:57]: 2 or 3?

**Review by JainTwinkle — APPROVED** [review]
LGTM :)

**Pull request merged**